chore(docker): update Docker image to support app development

jkuester opened the pull request:
Updates the existing `Dockerfile`:

- Use Node 16 (from the official Node Docker image)
- Include the various libraries necessary for running tests with the `cht-conf-test-harness` (e.g. `xsltproc`)
- Set `cht` command as the ENTRYPOINT (to allow for streamlined command execution)
- Run with non-root user

Adds documentation for how to use the image as a VSCode Dev Container or as a standalone Docker utility.

Closes #542 


## Not included

The following are not included (yet) in this MVP PR.  Part of this review should be deciding if any of these are necessary for merging this PR (or if they can be added later).  

- [ ] Automatically build/publish the Docker image. Currently to use the Docker image you have to clone the `cht-conf` repo and build the image yourself.  These steps would be unnecessary if we had GH Action steps that would build/deploy images publicly.
- [ ] Better support for connecting to locally hosted Docker instances.  Currently the documentation suggests connecting using the local IP of the host system and the `--accept-self-signed-certs` flag.  There are probably better ways to do this.
  - Edit, I realized that `my.local-ip.co` addresses should work fine as long as the port is open, so perhaps this is not too big of a deal.  If someone is actually running a development instance of the CHT on their bare-metal local machine, they probably already have the dependencies they need to use cht-conf....

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

jkuester commented on Dockerfile:
```diff
@@ -1,23 +1,28 @@
-FROM ubuntu:18.04
-MAINTAINER DevOps "devops@medicmobile.org"
+FROM node:16-slim
```
The ultimate container size is `~1GB`.  I spent a considerable amount of time trying to get things to work with an `alpine` container, but never could get it to work with the way that our config tests run chromium via the test harness and Puppeteer.  

Also I figured that for a VSCode Dev Container setup Debian might be a better environment than Alpine (more familiar to someone using a terminal and possibly more compatible [Dev Container Features](https://containers.dev/features)).

jkuester commented:
FYI this was developed and tested on Linux, but I also have tested that this process works on Windows too!  Using WSL, I was able to build the Docker image and then VSCode seamlessly picked up the image and everything worked as expected.

mrjones-plip commented:
@jkuester  - good stuff! I think we're close, but I think the docs in the readme need help.  I got stuck because I've literally never used VSCode to write code (just as a nice text editor for .txt files).  I didn't really know how to create a workspace and when I thought I did, and added the `.devcontainer.json` file, choosing `Reopen in Container` didn't work - instead t was prompting me to define a container or something? Video below.

I think we should be prescriptive toward a CHT experience - adding a new CHT app project with app settings initialized etc. And, yes, add a pre-built and published image (5 min build time for me both times!).

I'll grab ya tomorrow and we'll figure it out!

https://user-images.githubusercontent.com/8253488/222061857-0228fc86-fba7-4e98-871f-c76f7fbd44f5.mp4

andrablaj commented:
> FYI this was developed and tested on Linux, but I also have tested that this process works on Windows too! Using WSL, I was able to build the Docker image and then VSCode seamlessly picked up the image and everything worked as expected.

FYI, I just tested the `README` steps in MacOS with M1 chip, and everything ran correctly. 🎆 

The only hiccup I encountered was when using VSCode's integrated terminal, where I got the error below. I didn't investigate further, and just used my Mac's terminal instead.

<img width="708" alt="Screenshot 2023-03-01 at 12 58 18" src="https://user-images.githubusercontent.com/24651706/222146079-2c6cbb37-9215-4cc0-96bb-a55c250e095c.png">

jkuester commented:
For the record, the problem encountered by @mrjones-plip as due to his `.devcontainer.json` being located at `./test/.devcontainer.json` and not in the root of the workspace at `./.devcontainer.json`.

Also, @andrablaj, your error is coming because you are trying to mix the usage modes!  :smile: When using VSCode's integrated terminal (while having VSCode open in the Dev Container) the terminal is running _in the container environment_ and not _in your host environment_.  So, even though you have docker installed on your host machine, you cannot use the `docker` command from VSCode's integrated terminal because Docker is not installed _inside the container_!  To run `cht` commands from within VSCode, you don't need the extra `docker run ...` stuff. You can just run `cht ...`.  

I will update the documentation here to be more clear on all these points!

andrablaj commented:
Thanks @jkuester. 🙌  An update to the documentation would help, as more people might bump into the same question. I am stepping out of this PR, unless you need extra testing on MacOS. If it is the case, just ping me, I am happy to help.

mrjones-plip commented on README.md:
````diff
@@ -32,22 +32,66 @@ As Administrator:
+
+[Install VSCode](https://code.visualstudio.com/) if you do not have it already.
+
+Run the following commands from within your project directory to download the `.devcontainer.json` config file, install the [Dev Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), and open the project directory in VSCode:
+
+```shell
+curl https://raw.githubusercontent.com/medic/cht-conf/main/devcontainer.config/.devcontainer.json > .devcontainer.json
+code --install-extension ms-vscode-remote.remote-containers
+code -n .
````
Let's assume that if they have an existing project directory, they'll know how to use it.  But that if they don't, they'll need a lot of help, and so default to creating it for them. As well, let's quiet up `curl`. Finally let's call out they need to use the `wsl` shell not the `cmd` shell:

```suggestion
Using a WSL shell (Start -> wsl) to run the following commands. It will create a new project directory, download the `.devcontainer.json` config file, install the [Dev Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), and open the project directory in VSCode:

```shell
mkdir -p ~/cht-vscode-project && cd  ~/cht-vscode-project
curl -s https://raw.githubusercontent.com/medic/cht-conf/main/devcontainer.config/.devcontainer.json > .devcontainer.json
code --install-extension ms-vscode-remote.remote-containers
code -n .
```

jkuester replied:
So, they should have already created the project directory above (when building the Docker image). Added a note here to clarify...

Also, I think it is a good idea to mention WSL for Windows, but I hesitate to make that the only language we use since anecdotally a lot of app developers use non-Windows systems too.

mrjones-plip replied:
ah - dude - why am I only thinking of WSL?!  That's silly of me and you're right - so many OSes out there!  your mention of WSL is great.  I'll resolve this.

mrjones-plip commented on README.md:
```diff
@@ -32,22 +32,66 @@ As Administrator:
-	docker stop cht-conf
+If you want to develop CHT apps with VSCode, you can use the Docker image as a Development Container. This will allow you to use the `cht-conf` utility and its associated tech stack from within VSCode (without needing to install dependencies like NodeJS on your host system).
+
+[Install VSCode](https://code.visualstudio.com/) if you do not have it already.
```
```suggestion
Install both [VSCode](https://code.visualstudio.com/) and [Windows Subsystem for Linux (WSL2)](https://learn.microsoft.com/en-us/windows/wsl/install) if you haven't already.
```

mrjones-plip commented on README.md:
````diff
@@ -32,22 +32,66 @@ As Administrator:
+code --install-extension ms-vscode-remote.remote-containers
+code -n .
+```
+
````
I got a scary warning that feels like we should account for:

```suggestion

If this is the first time you're running VSCode in this directory, you may get a prompt:

> **Do you trust the authors of the files in this folder?**

Choose, "Yes, I trust the authors".

```

mrjones-plip reviewed (changes requested):
Yes!  This is looking really good.  

Catering to the first time dev, I'm requesting some more hand holding content be added.

mrjones-plip commented on README.md:
````diff
@@ -32,22 +32,66 @@ As Administrator:
+code -n .
+```
+
+Open the Command Palette in VSCode (_Ctrl+Shift+P_ or _Cmd+Shift+P_) and select `Reopen in Container`. This will open your workspace inside the `cht-conf` container. You can use the `cht` commands by opening a terminal in VSCode (_Ctrl+Shift+\`_ or _Cmd+Shift+\`_).
````
```suggestion
Open the Command Palette in VSCode (_Ctrl+Shift+P_ or _Cmd+Shift+P_) and select `Reopen in Container`. This will open your workspace inside the `cht-conf` container. You can use the `cht` commands by opening a terminal in VSCode (_Ctrl+Shift+\`_ or _Cmd+Shift+\`_). If prompted "Do you trust the authors..." choose "Trust Folder & Continue".
```

mrjones-plip replied:
I got a second prompt!

jkuester commented:
@Hareet I noticed that you were the one who originally created the Dockerfile here in cht-conf.  Before we haul-off and make fundamental changes here, I wanted to check and see if you all were using this for some kind of scripting somewhere that we should be aware of (so we don't inadvertently break anything)?

jkuester commented:
@Hareet I noticed that you were the one who originally created the Dockerfile here in cht-conf.  Before we haul-off and make fundamental changes here, I wanted to check and see if you all were using this for some kind of scripting somewhere that we should be aware of (so we don't inadvertently break anything)?

mrjones-plip commented on README.md:
```diff
@@ -32,22 +32,66 @@ As Administrator:
+
+Open the Command Palette in VSCode (_Ctrl+Shift+P_ or _Cmd+Shift+P_) and select `Reopen in Container`. This will open your workspace inside the `cht-conf` container. You can use the `cht` commands by opening a terminal in VSCode (_Ctrl+Shift+\`_ or _Cmd+Shift+\`_).
+
+Run the following command in the VSCode terminal to bootstrap your new CHT project:
```
```suggestion
In VSCode type `ctrl` + `` ` ``  to open a terminal in your Dev Container, and run the following command to bootstrap your new CHT project:
```

jkuester replied:
This feels unnecessary since we just told them how to open the terminal on the previous line...  :thinking:

mrjones-plip replied:
oh - i didn't rtfm - sorry!  Agreed, redundant, I'll self resolve this.

mrjones-plip reviewed (changes requested):
Came back in with a few more suggestions - still lookin' good!

mrjones-plip commented on README.md:
````diff
@@ -32,22 +32,74 @@ As Administrator:
+[Install VSCode](https://code.visualstudio.com/) if you do not have it already.
+
+Using the terminal (or the WLS shell on Windows: _Start > wsl_), run the following commands from within your project directory (created above) to download the `.devcontainer.json` config file, install the [Dev Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), and open the project directory in VSCode:
+
+```shell
+curl https://raw.githubusercontent.com/medic/cht-conf/main/devcontainer.config/.devcontainer.json > .devcontainer.json
+code --install-extension ms-vscode-remote.remote-containers
+code -n .
+```
````
Minor point to keep the whole "copy paste and it just works!" vibe goin by adding a `cd` and then still adding a `-s` to quiet up `curl`:

```suggestion

```shell
cd ~/cht-project
curl -s https://raw.githubusercontent.com/medic/cht-conf/main/devcontainer.config/.devcontainer.json > .devcontainer.json
code --install-extension ms-vscode-remote.remote-containers
code -n .
```
```

mrjones-plip reviewed (approved):
yay!  Thanks for dealing my absent mindedness on my review - all looks good.

minor suggestion made on a command, but approving to unblock let you decide how to proceed.

mrjones-plip commented on README.md:
````diff
@@ -32,22 +32,74 @@ As Administrator:
+[Install VSCode](https://code.visualstudio.com/) if you do not have it already.
+
+Using the terminal (or the WLS shell on Windows: _Start > wsl_), run the following commands from within your project directory (created above) to download the `.devcontainer.json` config file, install the [Dev Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), and open the project directory in VSCode:
+
+```shell
+cd ~/cht-project
+curl -s https://raw.githubusercontent.com/medic/cht-conf/main/devcontainer.config/.devcontainer.json > .devcontainer.json
+code --install-extension ms-vscode-remote.remote-containers
+code -n .
+
````
oops! I think my suggestion last time ate the ` ``` ` needed on line 67.  I can't figure out how to make a suggestion  that has it in there, but it looks link this now:

![image](https://user-images.githubusercontent.com/8253488/223195581-597e7b40-cc2d-4f7f-a462-143d3ac7069a.png)

jkuester replied:
Oh good catch!  Pushed a fix.

mrjones-plip reviewed (changes requested):
yikes - sorry! I think my suggestion caused one set of ` ``` ` to not be there.

jkuester commented:
@mrjones-plip I made a few minor updates here:

- Switched to use `cht-ide` wording for consistency with the upcoming published Docker image
- Added the `python3-wheel` package to the Dockerfile to resolve an error that was being printed when installing pyxform (the old Dockerfile had the package, but I removed it because I did not understand what it was for. Putting it back now because I saw the error....)

mrjones-plip reviewed (approved):
Looks good - thanks!
